src/shogun/lib/auto_initialiser.h

```diff
@@ -29,36 +32,102 @@ namespace shogun
 
 			~GammaFeatureNumberInit() override = default;
 
-			Any operator()() override
+			Any operator()() final
 			{
-				REQUIRE(m_kernel != nullptr, "m_kernel is not pointing to a CKernel object");
+				REQUIRE(m_kernel != nullptr, "No kernel provided!");
 
 				auto lhs = m_kernel->get_lhs();
 				switch (lhs->get_feature_class())
 				{
-					case EFeatureClass::C_DENSE:
-					case EFeatureClass::C_SPARSE:
+				case EFeatureClass::C_DENSE:
+				{
+					auto dense_features = (CDenseFeatures<float64_t>*)lhs;
+					return make_any(
+					    1.0 /
+					    static_cast<float64_t>(
+					        dense_features->get_dim_feature_space()) *
+					    dense_features->std(false)[0]);
+				}
+				case EFeatureClass::C_SPARSE:
+				{
+					auto dot_features = (CDotFeatures*)lhs;
+					return make_any(
+					    1.0 / (static_cast<float64_t>(
+					               dot_features->get_dim_feature_space()) *
+					           dot_features->get_std(false)[0]));
+				}
+				default:
+					auto dot_features = (CDotFeatures*)lhs;
+					return make_any(
+					    1.0 / static_cast<float64_t>(
+					              dot_features->get_dim_feature_space()));
+				}
+			}
+
+		private:
+			CKernel* m_kernel;
+			SG_DELETE_COPY_AND_ASSIGN(GammaFeatureNumberInit);
+		};
+
+		class GaussianWidthAutoInit : public AutoInit
+		{
+			static const char* const kName;
+			static const char* const kDescription;
+
+		public:
+			explicit GaussianWidthAutoInit(CKernel* kernel)
+			    : AutoInit(kName, kDescription), m_kernel(kernel)
+			{
+			}
+
+			~GaussianWidthAutoInit() override = default;
+
+			Any operator()() final
+			{
+				REQUIRE(m_kernel != nullptr, "No kernel provided!");
+				auto lhs = m_kernel->get_lhs();
+				auto rhs = m_kernel->get_rhs();
+				switch (lhs->get_feature_class())
+				{
+				case EFeatureClass::C_DENSE:
+				{
+					auto distance_kernel = CEuclideanDistance();
+					distance_kernel.init(lhs, rhs);
+					auto pdist = distance_kernel.get_distance_matrix();
+					auto result = SGVector<float64_t>(
+					    (lhs->get_num_vectors() * lhs->get_num_vectors() -
+					     lhs->get_num_vectors()) /
+					    2);
+					// copy upper triangular wihout a particular order
+					index_t idx = 0;
+					for (int i = 0; i < lhs->get_num_vectors(); ++i)
 					{
-						auto dot_features = (CDotFeatures*)lhs;
-						return make_any(
-						    1.0 /
-						    (static_cast<float64_t>(
-								dot_features->get_dim_feature_space()) *
-								dot_features->get_std(false)[0]));
+						for (int j = i + 1; j < lhs->get_num_vectors(); ++j)
```

Comment:
It would be cool if we could not have explicit loops for traversing triangular matrices everywhere, but rather do this via a utility iterator or something. See comment on the pdist as vector above